pkg/reconciler/taskrun/taskrun.go

```diff
@@ -228,27 +230,15 @@ func (c *Reconciler) checkPodFailed(tr *v1beta1.TaskRun) (bool, v1beta1.TaskRunR
 	return false, "", ""
 }
 
-func (c *Reconciler) durationAndCountMetrics(ctx context.Context, tr *v1beta1.TaskRun) {
+func (c *Reconciler) durationAndCountMetrics(ctx context.Context, tr *v1beta1.TaskRun, before *apis.Condition) {
 	logger := logging.FromContext(ctx)
 	if tr.IsDone() {
-		newTr, err := c.taskRunLister.TaskRuns(tr.Namespace).Get(tr.Name)
-		if err != nil && !k8serrors.IsNotFound(err) {
-			logger.Errorf("Error getting TaskRun %s when updating metrics: %w", tr.Name, err)
-			return
-		} else if k8serrors.IsNotFound(err) || newTr == nil {
-			logger.Debugf("TaskRun %s not found when updating metrics: %w", tr.Name, err)
-			return
-		}
-
-		before := newTr.Status.GetCondition(apis.ConditionSucceeded)
-		go func(metrics *taskrunmetrics.Recorder) {
-			if err := metrics.DurationAndCount(ctx, tr, before); err != nil {
-				logger.Warnf("Failed to log the metrics : %v", err)
-			}
-			if err := metrics.CloudEvents(ctx, tr); err != nil {
-				logger.Warnf("Failed to log the metrics : %v", err)
-			}
-		}(c.metrics)
+		if err := c.metrics.DurationAndCount(ctx, tr, before); err != nil {
+			logger.Warnf("Failed to log the metrics : %v", err)
+		}
+		if err := c.metrics.CloudEvents(ctx, tr); err != nil {
+			logger.Warnf("Failed to log the metrics : %v", err)
+		}
```

Comment:
By reading this change, since we pass the `tr` to this function, and the `before` is got from the `tr`, why don't we just pass `tr` and inside the function get the `before` there?

Is it because the `before` may change later in `tr`? since `durationAndCountMetrics` is deferred. If so, in previous code we fetch the tr again from the lister, so it should be the latest `tr` not the previous `tr`. This seems to be a behaviour change?

Reply:
Thanks for the detailed review Yongxuan! Yes you are right, tr will be modified later on in this reconcile loop. It's worth noticing that though the value of this tr is changed, but it hasn't been saved into the etcd, therefore the condition retrieved from the lister is the same as the "before" condtion we get from tr initially. does this make sense to you?

Reply:
yes that makes sense! I think the PR's approach is good. 
Just one question, I read Andrea's comment and I don't think lister will talk to api server/etcd, it is a cache and shouldn't fetch data from etcd?

Reply:
ref: https://aly.arriqaaq.com/kubernetes-informers/

Reply:
Yes yes, you are right, I believe so, it is fetching data from the shared informer instead of from the API server (which makes it less costy). Nevetheless, this is still an unnecessary api call, while every API call that eats network bandwith. Though it looks nother for one API call, but that's a big difference if we schedule taskruns at large scale :)

Reply:
Anyway I think we are aligned, thanks for clarifying the technical detail! Appreciate your expertise here!

Reply:
It should be a local storage and we just fetch the data from cache, my understanding is that it doesn't send request to api server.  What was the api call before (sorry I may miss that)?  Did you mean 
```
c.taskRunLister.TaskRuns(tr.Namespace).Get(tr.Name)
```

Reply:
Oh true, sorry I was being curtly on "network bandwith". The point I wantted to make is this is an unnecessary "call" :P

Reply:
sounds good! Shall we change the commit msg and pr description?

Reply:
SG! updated.